### PR TITLE
PR: assembly parser io ops tests + parse_mem implementation

### PR DESCRIPTION
**Pull request by grjte**
Hi Bobbin,

I added some tests to get my bearings. Please let me know if you prefer unit tests more or less collapsed (i.e. all mem variants in one test or splitting the 2 methods of providing address into their own tests).

Hopefully I'm on the right track here!

### CODE REVIEW

File: assembly/src/v1/parsers/io_ops.rs [hunk 1]
```diff
@@ -50,7 +50,218 @@ pub fn parse_readw(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), As
...
+/// load: reads a word from memory and overwrites the top 4 elements of the stack. If an address was
+/// provided, PUSH the address, then LOADW to load a word from memory.
+///
+/// pop: pops the top 4 elements off the stack and saves them to memory. If an address was provided,
+/// PUSH the address first, STOREW to store the top 4 elements (below the address) at the specified memory address,
+/// then DROP 4 times to pop the elements off the stack
+///
+/// store: saves the top 4 elements of the stack to memory and leaves them on the stack. If an address was
+/// provided, PUSH the address first, then STOREW to save the top 4 elements to the specified memory location.
+pub fn parse_mem(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
```
Comment (bobbinth) [inline assembly/src/v1/parsers/io_ops.rs:70]: nit: I usually try to keep lines width within 100 chars. also, some sentences seem to be missing periods.

File: assembly/src/v1/parsers/io_ops.rs [hunk 2]
```diff
@@ -50,7 +50,218 @@ pub fn parse_readw(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), As
...
+            // push the address for LoadW
+            span_ops.push(Operation::LoadW);
+        }
+        "load" => span_ops.push(Operation::LoadW),
+        _ => return Err(AssemblyError::invalid_op(op)),
+    }
+
+    Ok(())
+}
+
```
Comment (bobbinth) [inline assembly/src/v1/parsers/io_ops.rs:122]: nit: maybe add separator lines here? I usually try to visually separate actual code from test code. Another option is to break out tests into a separate file.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Great job! (especially with tests). I left a couple of minor nits in-line.

One structural comment I have: I wonder if it would be better to break out handling of loads and stores into separate functions. It would add some overhead, but might improve code comprehension.

For example, we could have two helper functions (names could be different):
* `parse_mem_read()` which would handle processing of `mem.load` and `mem.push`
* `parse_mem_write()` which would handle processing of `mem.store` and `mem.pop`

What do you think?

**Comment by grjte** [comment]
Yes, I think it's clearer what's happening with it broken out into read/write functions. The most confusing op is still mem.push, but hopefully it's all more readable now.

File: assembly/src/v1/parsers/io_ops.rs (hunk 1 above)
Comment (grjte) [reply assembly/src/v1/parsers/io_ops.rs]: Fixed.

File: assembly/src/v1/parsers/io_ops.rs (hunk 2 above)
Comment (grjte) [reply assembly/src/v1/parsers/io_ops.rs:141]: Added a separator line, since it seems to be idiomatic in Rust to keep unit tests in the same file.

**Review by bobbinth — APPROVED** [review]
Looks great! Thank you!

**Pull request merged**